### PR TITLE
[SoX with conflict checking] Check time model for transaction submissions

### PR DESCRIPTION
**Pull request by tudor-da**
This PR implements time model check for transaction submissions.
Additionally, some QoL refactoring (see justification as review comments).

**NOTE** There exists no conformance test that asserts this check. Unit tests are to be implemented in a follow-up PR.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala [hunk 1]
```diff
@@ -132,22 +133,22 @@ private[sandbox] class ConflictCheckingLedgerBridge(
         Timed
           .future(
             bridgeMetrics.Stages.conflictCheckWithCommitted,
-            checkTimeModel(originalSubmission) match {
```
Comment (tudor-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala:135]: `checkTimeModel` is performed in `sequence` in order to have access to the record time.

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala [hunk 2]
```diff
@@ -377,56 +378,49 @@ private[sandbox] class ConflictCheckingLedgerBridge(
...
               }
-              .map(
-                _.left.map(toCommandRejectedUpdate(_, transaction.submitterInfo.toCompletionInfo()))
-              )
           case left => Future.successful(left)
         }
     }
+  }
 
-  private def toCommandRejectedUpdate(rejection: Rejection, completionInfo: CompletionInfo) =
```
Comment (tudor-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala:395]: Moved to `Rejection`

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala [hunk 3]
```diff
@@ -377,56 +378,49 @@ private[sandbox] class ConflictCheckingLedgerBridge(
...
-      completionInfo = completionInfo,
-      reasonTemplate = FinalReason(rejection.toStatus),
-    )
-
-  private def invalidInputFromParticipantRejection(
-      completionInfo: CompletionInfo
-  )(implicit
+  private def invalidInputFromParticipantRejection(completionInfo: CompletionInfo)(implicit
       contextualizedErrorLogger: ContextualizedErrorLogger
-  ): LfTransaction.KeyInputError => Update.CommandRejected = {
```
Comment (tudor-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala:406]: `recordTime` is captured in `sequence` together with offsets assignment. Then, let's generate plain rejections on earlier stages and generate the CommandRejected as the last step in the `sequence` stage.

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/domain/Rejection.scala [hunk 4]
```diff
@@ -12,17 +14,27 @@ import lf.data.Time.Timestamp
...
 import platform.server.api.validation.ErrorFactories
 import platform.store.appendonlydao.events.ContractId
-
 import com.google.protobuf.any.Any
 import com.google.rpc.error_details.ErrorInfo
 import com.google.rpc.status.Status
 
 private[sandbox] sealed trait Rejection extends Product with Serializable {
   def toStatus: Status
+  def completionInfo: CompletionInfo
```
Comment (tudor-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/domain/Rejection.scala:23]: Add the completionInfo so we can generate `Update.CommandRejected` from `Rejection` without the need to thread `CompletionInfo` through by different means

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala [hunk 5]
```diff
@@ -293,20 +300,27 @@ private[sandbox] class ConflictCheckingLedgerBridge(
...
+      .toRight(Rejection.NoLedgerConfiguration(completionInfo, errorFactories))
+      .flatMap(configuration =>
+        configuration.timeModel
+          .checkTime(
+            transaction.transactionMeta.ledgerEffectiveTime,
+            recordTime,
+          )
+          .left
+          .map(Rejection.InvalidLedgerTime(completionInfo, _)(errorFactories))
+      )
```
Comment (tudor-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala:323]: This is the time check implementation.

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala [hunk 6]
```diff
@@ -87,18 +85,19 @@ private[sandbox] class ConflictCheckingLedgerBridge(
             .left
             .map(
               withErrorLogger(submitterInfo.submissionId)(
-                invalidInputFromParticipantRejection(submitterInfo.toCompletionInfo())(_)
+                invalidInputFromParticipantRejection(
+                  transactionSubmission.submitterInfo.toCompletionInfo()
```
Comment (mziolekda) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala:89]:
```suggestion
                  submitterInfo.toCompletionInfo()
```

**Review by mziolekda — APPROVED** [review]

**Pull request merged**